src/compiler/mir/ir.rs

```diff
@@ -0,0 +1,370 @@
+            temps: vec![],
+            span,
+        }
+    }
+
+    /// Add an argument to this procedure
+    pub fn add_arg(&mut self, name: StringId, ty: &Type) -> VarId {
+        let vd = VarDecl::new(name, false, ty, ScopeId::root());
+        self.vars.push(vd);
+        let id = self.vars.len() - 1;
+        VarId::new(id)
+    }
+
+    /// Get a [`BasicBlock`] for this procedure
+    pub fn get_bb(&self, id: BasicBlockId) -> &BasicBlock {
+        &self.blocks[id.index()]
+    }
+
+    /// Get a mutable [`BasicBlock`] from this procedure.
+    pub fn get_bb_mut(&mut self, id: BasicBlockId) -> &mut BasicBlock {
+        &mut self.blocks[id.index()]
+    }
+
+    /// Create a new, empty, [`BasicBlock`] in this procedure. This will
+    /// return the ID of that [`BasicBlock`].
+    pub fn new_bb(&mut self) -> BasicBlockId {
+        let bb = BasicBlock::new();
+        self.blocks.push(bb);
+        let id = self.blocks.len() - 1;
+        BasicBlockId::new(id)
+    }
+
+    /// Get the declaration details of a user defined variable in
+    /// this procedure.
+    pub fn get_var(&self, id: VarId) -> &VarDecl {
+        &self.vars[id.index()]
+    }
+
+    /// Get the declaration details of a temporary variable.
+    pub fn get_temp(&self, id: TempId) -> &TempDecl {
+        &self.temps[id.index()]
+    }
+
+    /// Add a new user defined variable to the procedure.
+    pub fn add_var(&mut self, name: StringId, mutable: bool, ty: &Type, scope: ScopeId) -> VarId {
+        let vd = VarDecl::new(name, mutable, ty, scope);
+        self.vars.push(vd);
+        let id = self.vars.len() - 1;
+
+        VarId::new(id)
+    }
+
+    /// Add a new temporary variable to the procedures stack
+    pub fn add_temp(&mut self, ty: &Type) -> TempId {
+        let td = TempDecl::new(ty);
+        self.temps.push(td);
+        let id = self.temps.len() - 1;
+        TempId::new(id)
+    }
+}
+
+/// Identifier for a specific basic block in a procedure
+#[derive(Debug, PartialEq, Copy, Clone)]
+struct BasicBlockId(usize);
+
+impl BasicBlockId {
+    pub fn new(id: usize) -> BasicBlockId {
+        BasicBlockId(id)
+    }
+
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Identifier for a user declared variable
+#[derive(Debug, PartialEq, Clone)]
+struct VarId(usize);
+
+impl VarId {
+    pub fn new(id: usize) -> VarId {
+        VarId(id)
+    }
+
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Identifier for a temporary variable.
+#[derive(Debug, PartialEq, Clone)]
+struct TempId(usize);
+
+impl TempId {
+    pub fn new(id: usize) -> TempId {
+        TempId(id)
+    }
+
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Identifier for each scope that exists within the function
+#[derive(Debug, PartialEq, Clone)]
+struct ScopeId(usize);
+
+impl ScopeId {
+    pub fn new(id: usize) -> ScopeId {
+        ScopeId(id)
+    }
+
+    pub fn root() -> ScopeId {
+        ScopeId::new(0)
+    }
```

Comment:
I think that it's a bad idea to have the `ScopeId` manage what the ID of the root scope is? It should be up to the either the Procedure or whatever manages the scope tree of a Procedure to determine what scope IDs are associated with specific scopes in a procedure.